Volha Halynskaya

Vaires opened the pull request:
Знаю, знаю что криво написано. Но все тесты на локальном прошли, это просто попытка показать что работает. Теперь можно и логику переворачивать делая всё более красивым и правильным.

Vaires committed:
First try to check calculator.

Vaires committed:
Update packages

Vaires committed:
Update setup.py

Vaires committed:
All test's fully passed.

Vaires committed:
Trying to fix setup.py

AlexeiBuzuma commented on final_task/my_tests.py:
```diff
@@ -0,0 +1,94 @@
+        self.assertEqual(pycalc.counting("1---1"), eval("1---1"), "FAILED")
+        self.assertEqual(pycalc.counting("-+---+-1"), eval("-+---+-1"), "FAILED")
+
+        # Operation priority
```
Как минимум эти группы проверок нужно разделить на разные тест кейсы

Vaires replied:
Не совсем понимаю что это значит. Проверки и их группы были взяты из файла которым с нами поделились в slack в самом начале.

Vaires replied:
Разделила их

AlexeiBuzuma commented on final_task/my_tests.py:
```diff
@@ -0,0 +1,94 @@
+import unittest
+import pycalc
+import math
+from math import *
```
Нужно избегать импортирование через звездочку

Vaires replied:
Сделано - провела импорт руками только функций участвующих в тест-кейсах

AlexeiBuzuma commented on final_task/pycalc.py:
```diff
@@ -0,0 +1,242 @@
+arg_count = 0
+
+
+def counting(expression):
```
Для чего используется эта функция? Все, что она делает -- это хранит другие функции

Vaires replied:
Сделано - вырезала. Изначально хотелось объединить всё в 1 функцию для упрощения понимания собственного т.к. функции появлялись постепенно по мере необходимости.

AlexeiBuzuma commented on final_task/pycalc.py:
```diff
@@ -0,0 +1,242 @@
+    '%': (3, "left", operator.mod), 'abs': (5, "left", operator.abs), '-abs': (5, "left", operator.abs),
+    'round': (5, "left", lambda x: round(x)), '-round': (5, "left", lambda x: round(x))
+}
+NEW_DICTIONARY = dict(list(OPERATORS.items()) + list(MATH_FUNCTIONS.items()) + list(MATH_FUNCTIONS_NEG.items()))
```
Странное название контсанты. Оно не описывает содержание этого словаря.

Vaires replied:
Сделано - переименовала название словаря в OPERATORS_MATH_FUNCTIONS

AlexeiBuzuma commented on final_task/pycalc.py:
```diff
@@ -0,0 +1,242 @@
+                yield op.string[:end]
+                del_index = end - 1
+
+    def shunting_yard(parsed_expression):
```
ПОчему эта функция оформлена в виде генератора?

Vaires replied:
Переделала, теперь это 2 списка и гораздо проще к восприятию. Но теперь еще приходится считать количество аргументов переданных функции и заносить его в стэк.

AlexeiBuzuma commented on final_task/pycalc.py:
```diff
@@ -0,0 +1,242 @@
+        for token in parsed_expression:
+            if token in MATH_FUNCTIONS:
+                stack.append(token)
+                global arg_count
```
если нужно хранить состояние, то лучше использовать класс, чем глобальные переменные

Vaires replied:
Убрала, я теперь не храню его. Количество аргументов польской нотации передается в стэк и читается оттуда в момент чтения функции

AlexeiBuzuma commented on final_task/pycalc.py:
```diff
@@ -0,0 +1,242 @@
+            else:
+                return stack[0]
+        except IndexError:
+            return mistake_error
```
В Python есть механизм исключений. Нет смысла использовать возращаемые значения для передачи информации об ошибке, когда для этого есть механизм исключений

Vaires replied:
Не совсем поняла в чем конкретно ошибка, но исправила на чистое исключение.

AlexeiBuzuma commented on final_task/pycalc.py:
```diff
@@ -0,0 +1,242 @@
+        global quantity_arg
+        quantity_arg = 0
+        if tokens_operator.__name__ == '<lambda>':
+            quantity_arg = tokens_operator.__code__.co_argcount
```
У функции может быть переменное количество аргументов (`*args, **kwargs`)
Это решение будет работать не всегда

Vaires replied:
Переписала своё решение, убрала функцию с подсчетом параметров. Теперь их список передается непосредственно в момент анализа параметров внутри скобок, их может быть столько сколько нужно функции.

AlexeiBuzuma commented on final_task/pycalc.py:
```diff
@@ -0,0 +1,242 @@
+            quantity = 0
+            for index in cut_str:
+                if index == '(':
+                    quantity = quantity + 1
```
`quantity += 1`

Vaires replied:
Исправлено - забыла про это в данной функции.

AlexeiBuzuma commented on final_task/pycalc.py:
```diff
@@ -0,0 +1,242 @@
+    def check_entrence(check_expr, mistake=False):
+        check_expr = checklog(check_expr)
+        if check_expr is "":
+            return "ERROR: expression is empty."
```
Лучше использовать исключения

Vaires replied:
Исправлено - сделана попытка и исключение.

AlexeiBuzuma commented on final_task/pycalc.py:
```diff
@@ -0,0 +1,242 @@
+        space_between_operator_error = "ERROR: You have to delete space(s) between the operator(s)."
+        check = re.findall(r'(=\s+=)|(>\s+=)|(<\s+=)|(!\s+=)|(/\s+/)|(\w\s+\w)', check_expr)
+        if check:
+            return space_between_operator_error
```
Лучше использовать исключения

Vaires committed:
Исправлена часть замечаний при PR.

Убрала вызов *, разбила на подгруппы

Vaires committed:
Исправлена часть замечаний

Удалена избыточная функция, часть ручных проверок заменена на механизм исключений, заменено именование словаря

Vaires committed:
Переписан алгоритм сортировочной станции.

Переписала алгоритм сортировочной станции на функцию не использующую генератор. Убрала избыточный код подсчитывающий количество параметров функции, теперь не используется переменная arg_count, польская запись сама забирает то что считает нужным для вычисления из стека.

Vaires committed:
Добавлен подсчет количества аргументов для функций

И хотя травис сработал корректно, для некоторых функций неверно рассчитывалось количество аргументов. Переделала, теперь алгоритм сортировочной станции считает количество переменных для функции и заносит его в стэк, а затем читает для определения понимания сколько операндов ему взять из стэка. Есть мысли как сделать еще проще польскую нотацию.

Vaires committed:
Упрощен код внутри обратной польской записи

Переписала алгоритм обратной польской записи. Существенно упростила её, вырезала часть кода где шел анализ определения списка параметров. Теперь не используется функция с определением числа аргументов, теперь в момент сортировочной станции для всех функций идет числа параметров внутри скобок. Это сделано для подсчета *args , **kwargs

AlexeiBuzuma commented:
1) Сообщения коммитов в гите должны нести какой-то смысл изменения, которое было сделано в этом коммите. Сообщение `Update packages` не несет особого смысла. Сообщения на русском языке тоже не рекомендуется использовать
2) Название функции должно четко обозначать, что она делает. Название `calc` несет очень мало информации и без контекста очень сложно понять, что оно делает
3) В данном решении есть места, где текст ошибки является возвращаемым значением из функции. Для работы с ошибками в Python есть замечательный механизм исключений.
4) Наверное освноная проблема решения в том, что функции достаточно тяжело читаются человеком. Функции содержать большое количество логики. В таких больших функциях очень легко допустить ошибку. Разбиение на маленькие функции или классы обычно позволяет делать код более читаемым и менее подверженным случайным ошибкам.
5) Советую много практиковаться и постоянно писать код. Есть много ресурсов, где можно найти большое количество небольших и интересных заданий. Например https://www.codewars.com/

Vaires commented:
Спасибо за ревью! Всё будет учтено, а из ресурсов для практики мне очень понравился Sololearn.
